### ISSUE TITLE
New motion preditor testing

### ISSUE THREAD

**Issue by Jarcho** [Jarcho/d2dx#8]
I've been working on a new motion predictor. This is going to need some testing to see if anything goes wrong with it.

The main difference with this one is it hooks into the game to change the positions of everything before the game renders a frame. This fixes a whole bunch of issues that can't be fixed the old way and also make predictor compatible with all renderers.

Changes from the old version:
* Perspective mode is much smoother. All the regular d2 issues still apply (things sliding around relative to the floor being the biggest).
* Text labels no longer drift/jiggle. All the regular d2 issues with text label placement still apply.
* CPU use when the game is minimized has been fixed.
* CPU use in game and menus has been fixed.
* Character animations in the menu no longer run faster than normal.
* V-sync is disabled by default in favour of a frame limiter. It can be enabled if you really want it.
* Support all versions starting from v1.00 except
  * 1.04
  * 1.09c
  * 1.10b (beta release)
  * 1.10s (beta release)
  * 1.13a (beta release)
  * 1.13b (beta release)
  * 1.14a
  * 1.14b

Known Issues:
* ~Multiplayer motion smoothing is not implemented~ (fixed 2023-05-26)
* ~Teleported units will be 'smoothly' moved to their new location~ (fixed 2023-05-19)
* Game crashes with other fps patches (e.g. Project Diablo 2). Those will need to be disables for this to work. (workaround added in 2023-08-08)

For variable refresh rate monitors I'm unsure if the refresh rate is detected correctly or not. You can force a specific frame rate in `d2fps.ini`.

Download:
[d2dx-test-2023-08-08.zip](https://github.com/Jarcho/d2dx/files/12296674/d2dx-test-2023-08-08.zip)

**Comment by weili0677** [comment]
this test is can't play games normally。
in menu is good，but when i in game the picture was stop（like FPS is 0）。even in single or multi player
and can you add “show fps” function？

**Comment by weili0677** [comment]
> this test is can't play games normally。 in menu is good，but when i in game the picture was stop（like FPS is 0）。even in single or multi player and can you add “show fps” function？

I copy glide3x.dll d2fps.dll and d2fps.ini and in game the fps is 0.with out d2fps.dll and d2fps.ini the fps like 24fps.

**Comment by Jarcho** [comment]
A couple of quick questions:
* Is the game actually not rendering frames? (You can check using any external fps tool. On Win10+ the xbox game bar will do this and is preinstalled)
* Do you have any other mods running?

**Comment by weili0677** [comment]
> A couple of quick questions:
> 
> * Is the game actually not rendering frames? (You can check using any external fps tool. On Win10+ the xbox game bar will do this and is preinstalled)
> * Do you have any other mods running?

1.about "show FPS",I got it.but inbuild "show fps" was easy to use.
2.I test in "Project Diablo 2" singleplayer.

**Comment by Jarcho** [comment]
Project Diablo 2 has it's own set of patches applied. There's probably a conflict between those and d2fps. I'll look into it at some point, but that will wait until after regular D2 is fully working.

**Comment by Jarcho** [comment]
New version has been uploaded. Can you run that and upload the log file when running Project Diablo 2?

**Comment by weili0677** [comment]
> New version has been uploaded. Can you run that and upload the log file when running Project Diablo 2?

https://drive.google.com/drive/folders/1QbU1iA_pWf3ZUupWPRk70wj1XnALmRs1?usp=sharing

OK,in PD2's mod,when I choose char into game,will crash c00000005
in regular D2 can into game,but when I move the char and the picture will black in a while(mean flash pictures between normal and black)

**Comment by Jarcho** [comment]
Can you also upload the log file for regular d2. (You can drag files into github's editor instead of using an external site)

Just to be clear the game is rendering, but you get a black screen only while the the character is moving?

You can try to enable vsync and see if that helps. Command line for that is `-dxvsync`. Otherwise I'll upload a new version with more logging and see if that helps narrow it down.

**Comment by weili0677** [comment]
> 

OK,enable vsync is also has issues.and I upload the logs and vedios.
[D2_1.13c_pd2_logsAndVideo.zip](https://github.com/Jarcho/d2dx/files/11512664/D2_1.13c_pd2_logsAndVideo.zip)
[D2_1.13c_regular_logsAndVideo.zip](https://github.com/Jarcho/d2dx/files/11512666/D2_1.13c_regular_logsAndVideo.zip)

**Comment by Jarcho** [comment]
Couple things to try:
* The maphack is possibly conflicting. Can you try regular d2 without it.
* If that doesn't fix it can you do a quick check with motion smoothing disabled (`enable_smoothing=false` in d2fps.ini). If that works that at least narrows down where the problem is.

**Comment by weili0677** [comment]
> Couple things to try:
> 
> * The maphack is possibly conflicting. Can you try regular d2 without it.
> * If that doesn't fix it can you do a quick check with motion smoothing disabled (`enable_smoothing=false` in d2fps.ini). If that works that at least narrows down where the problem is.

test result is:
no maphack+enable_smoothing=true      =>   no good.
no maphack+enable_smoothing=false     =>   good.
maphack+enable_smoothing=false          =>   good.

**Comment by Jarcho** [comment]
That's at least something to looks at. I still can't reproduce it, so I'm going to have to add some tracing and have you run a new build. Should be sometime tomorrow.

**Comment by weili0677** [comment]
> 

Ok,Take it easy.I can help you.

**Comment by Jarcho** [comment]
Can you run [this version](https://github.com/Jarcho/d2dx/files/11519028/d2dx-test-2023-05-19.zip) and upload the log file with regular d2. This will log the player's position as you move around.

**Comment by weili0677** [comment]
> Can you run [this version](https://github.com/Jarcho/d2dx/files/11519028/d2dx-test-2023-05-19.zip) and upload the log file with regular d2. This will log the player's position as you move around.

OK，I have busy now。will test it later。

**Comment by weili0677** [comment]
[D2_113c_regular_good.zip](https://github.com/Jarcho/d2dx/files/11522096/D2_113c_regular_good.zip)
I test the newest version is good for regular D2,but PD2's mod also got c000005 error.

**Comment by Jarcho** [comment]
I was not expecting the last change to have fixed it, but good that it did. PD2 compatibility will wait until I finish porting this to other versions.

**Comment by weili0677** [comment]
> I was not expecting the last change to have fixed it, but good that it did. PD2 compatibility will wait until I finish porting this to other versions.

OK，Thank you.

**Comment by pas-de-2** [comment]
Tested out d2dx-test-2023-05-17.zip with 1.09d and 1.10:
1. Whenever my character stops moving, it feels like the screen 'jerks' momentarily. Hard to explain, but it's unpleasant, almost like a desync.
2. The FPS limiter doesn't seem to work exactly right. If I set fps=120, the FPS is constantly bouncing around from ~123-131. If I set it to fps=60, the FPS hovers around ~61-68.

**Comment by pas-de-2** [comment]
Same behavior, [d2fps.log](https://github.com/Jarcho/d2dx/files/11537464/d2fps.log)

**Comment by Jarcho** [comment]
The log makes it look like whatever you're using to measure the frame rate is broken. It's making 60 draw calls per second.

Can you record a short video the what the movement looks like? I can guess what you're describing, but it would be easier to see it.

**Comment by pas-de-2** [comment]
I'm using the FPS counter from Steam overlay, so I'm not sure what's up. It's never given me inaccurate results before.

Here's a video of the issue, hopefully you can see what I'm talking about:


https://github.com/Jarcho/d2dx/assets/116680301/3049cd71-5be7-4da2-a851-477aefafdf94

**Comment by Jarcho** [comment]
Can you try [this version](https://github.com/Jarcho/d2dx/files/11537950/d2dx-test-2023-05-22-2.zip). It should less of the problem.

I just checked steam's fps counter and I get the correct value. Not sure what's going on with that. Can you try with a different one to confirm? (Win+G get the xbox game bar)

**Comment by pas-de-2** [comment]
That's definitely an improvement 👍

I tried the Xbox game bar and it gives me the same FPS reading, strange. Always averaging like 5% over the cap.

**Comment by Jarcho** [comment]
That makes even less sense. The menu and game use the same timer. [One more test](https://github.com/Jarcho/d2dx/files/11538866/d2dx-test-2023-05-22.3.zip). This time I'll need `d2dx_log.txt`.

At least the important issue was fixed.

**Comment by pas-de-2** [comment]
[d2dx_log.txt](https://github.com/Jarcho/d2dx/files/11539174/d2dx_log.txt)

**Comment by Jarcho** [comment]
You'll need to replace the `glide3x.dll`, not just `d2fps.dll`.

**Comment by pas-de-2** [comment]
I did! Double checked to be sure, the log output is identical. Here's the d2fps.log:
[d2fps.log](https://github.com/Jarcho/d2dx/files/11539589/d2fps.log)

**Comment by Jarcho** [comment]
From `d2dx-test-2023-05-22.3.zip`? That's not the right log output for either of those files.

**Comment by pas-de-2** [comment]
Oof, I could have sworn I downloaded that, was just going by the latest zip I had. Sorry!
[d2dx_log.txt](https://github.com/Jarcho/d2dx/files/11539651/d2dx_log.txt)

**Comment by Jarcho** [comment]
That's running the game at ~120 fps. There were a few fluctuations, but the average is just slightly under 120. Given that this is the amount of time passed between present calls there isn't a more accurate measurement of what the game is drawing. If you're getting a higher number then it's from something else.

**Comment by Jarcho** [comment]
Main post updated. Nearly all versions are now supported.

**Comment by Jarcho** [comment]
Checked into PD2 crashing. The crash is caused by PD2 patching over memory d2fps already patched. Fixing this would require PD2 to not apply their fps patches.

**Comment by Jarcho** [comment]
New release. Multiplayer should be "working" (as in doesn't crash), but needs testing for how motion smoothing actually behaves. Mainly needs testing around how motion smoothing combines with network latency.

**Comment by pas-de-2** [comment]
Here's a strange one. I'm getting assertion errors sometimes when I enter TPs from town. I've noticed this primarily happens when trying to TP back to the Outer Cloister/Barracks/Jail Lvl 1 area, but I've also had it happen once when trying to TP back to Arcane Sanctuary.

![hRoom](https://github.com/Jarcho/d2dx/assets/116680301/f7faf15c-96a9-44eb-ac3c-7ca41f8cb254)
It's difficult to reproduce reliably, but just leveling in the Outer Cloister and the Barracks as a Sorceress I ran into this several times. I couldn't reproduce this on the release version of D2DX. I did manage to replicate it on both 1.05b and 1.09d. 

Here's the D2 crash log if it's of any value: [D2230601.txt](https://github.com/Jarcho/d2dx/files/11626375/D2230601.txt)

**Comment by Jarcho** [comment]
Couple quick questions
* Single of multi player? (looks like single)
* Does the crash only happen when teleporting in one direction, or both?

**Comment by pas-de-2** [comment]
SP. Only happens when teleporting from town to somewhere else, not when teleporting back to town.

Here's my 1.09d sorc:
[109d-sorc.zip](https://github.com/Jarcho/d2dx/files/11628398/109d-sorc.zip)

**Comment by Jarcho** [comment]
Still can't get it with that. Can you upload the whole save and not just the character.

**Comment by pas-de-2** [comment]
Regenerating the map didn't make a difference for me: [109d-sorc2.zip](https://github.com/Jarcho/d2dx/files/11632050/109d-sorc2.zip)
>I assume you're using the latest version 2023-05-26.

Yes, but I also tried 5-22-2 and that crashed, too.

**Comment by Jarcho** [comment]
I have no idea how you're triggering that. Over a hundred TP's while running around and I still haven't gotten it to crash.

**Comment by Jarcho** [comment]
I guess it's the same issue that lead to me disabling smoothing in multiplayer originally then. I had never seen it crash single player though.

**Comment by Jarcho** [comment]
One last test version. Assuming there are no new problems I'll be updating the main release this weekend.

Thank you to everyone testing this.

**Comment by tomasz1986** [comment]
I don't know if you remember but I was the one with a lot of problems in https://github.com/bolrog/d2dx/pull/177 related to the background being blurry when moving. Right now I'm testing the newest version from here (https://github.com/Jarcho/d2dx/files/11742733/d2dx-test-2023-06-14.zip) and I must say that the blurry background issues seem to be gone 👍. I'm now going to try playing for a longer period of time to see whether everything else is stable as well. Thank you once again for all the work!

**Comment by Jarcho** [comment]
Good to hear it's working.

The black frame issue isn't new. I've had it happen with bolrog's last release as well as with all of mine. Haven't bothered to track down what's happening yet. Is it just flashing black, or is it a sustained black for a whole second?

I wouldn't recommend using vsync at all. It's essentially a bad implementation of a frame limiter with how it's implemented as it's done for throughput at the expense of latency.

Looks like you have the same issue @pas-de-2 with the frame limiter. I don't know what's happening there as d2dx is presenting the frames at the correct pacing, but for some reason it's being read as something slightly higher. Given that it doesn't happen to me there's zero chance I'm going to debug this.

Since you have an adaptive frame rate, does the auto-detection work correctly? You can set the frame rate to zero and check `d2fps.log` to see what it's detected as.

**Comment by tomasz1986** [comment]
> The black frame issue isn't new. I've had it happen with bolrog's last release as well as with all of mine. Haven't bothered to track down what's happening yet. Is it just flashing black, or is it a sustained black for a whole second?

It was more like a full second, and then the screen came back to normal. I was probably lucky not to experience it before then 🙂.

> Since you have an adaptive frame rate, does the auto-detection work correctly? You can set the frame rate to zero and check `d2fps.log` to see what it's detected as.

It does detect the refresh rate properly.

```
Attaching D2fps...
Error parsing d2fps.ini: line `1`: key with no value `;;; Config`
Error parsing d2fps.ini: line `3`: key with no value `; Force the game to run at a specific framerate`
Error parsing d2fps.ini: line `4`: key with no value `; Setting this to `0` will auto-detect the monitor's framerate`
Error parsing d2fps.ini: line `5`: unknown key `; CLI Argument: -fps`
Error parsing d2fps.ini: line `6`: key with no value `; Default: 0`
Error parsing d2fps.ini: line `9`: key with no value `; Force the game to run at a specific framerate when not the active window`
Error parsing d2fps.ini: line `10`: key with no value `; Setting this to `0` will use the foreground framerate.`
Error parsing d2fps.ini: line `11`: unknown key `; CLI Argument: -bg-fps`
Error parsing d2fps.ini: line `12`: key with no value `; Default: 25`
Error parsing d2fps.ini: line `15`: key with no value `;;; Features`
Error parsing d2fps.ini: line `17`: key with no value `; Whether or not to enable fps unlock and cpu use fix in the menu`
Error parsing d2fps.ini: line `18`: unknown key `; CLI Argument: -pmenu-fps`
Error parsing d2fps.ini: line `19`: key with no value `; Default: true`
Error parsing d2fps.ini: line `22`: key with no value `; Whether or not to enable fps unlock and cpu use fix in the game`
Error parsing d2fps.ini: line `23`: unknown key `; CLI Argument: -pgame-fps`
Error parsing d2fps.ini: line `24`: key with no value `; Default is true`
Error parsing d2fps.ini: line `27`: key with no value `; Whether or not to enable motion smoothing`
Error parsing d2fps.ini: line `28`: key with no value `; Requires: `game-fps``
Error parsing d2fps.ini: line `29`: unknown key `; CLI Argument: -pmotion-smoothing`
Error parsing d2fps.ini: line `30`: key with no value `; Default: true`
Loaded config:
  fps: 0
  bg-fps: 25
  features: [`menu fps`, `game fps`, `motion smoothing`]
Detected game version: 1.13c
Failed to apply patch at: D2Win.dll+0x18a4e
Disabling feature `menu fps`: failed to apply patches
Applied feature `game fps`
Applied feature `motion smoothing`
Loaded modules:
  Diablo 2\Game.exe (v1.0.13.60)
  Diablo 2\Storm.dll (v1999.11.18.1)
  Diablo 2\D2sound.dll
  Diablo 2\Fog.dll
  Diablo 2\D2Win.dll
  Diablo 2\D2MCPClient.dll
  Diablo 2\D2gfx.dll
  Diablo 2\DSOUND.dll (v0.20.0.0)
  Diablo 2\ijl11.dll (v1.1.2.16)
  Diablo 2\D2CMP.dll
  Diablo 2\D2Lang.dll
  Diablo 2\BaseMod.dll (v1.1.3.9)
  Diablo 2\D2Client.dll
  Diablo 2\D2Net.dll
  Diablo 2\D2Game.dll
  Diablo 2\D2Common.dll
  Diablo 2\BH.dll
  Diablo 2\D2MULTI.dll
  Diablo 2\D2LAUNCH.dll
  Diablo 2\Bnclient.dll
  Diablo 2\PlugY.dll (v14.0.3.0)
  Diablo 2\StormLib.dll
  Diablo 2\D2Glide.dll
  Diablo 2\glide3x.dll (v0.100.2.0)
  Diablo 2\smackw32.dll (v3.0.0.0)
  Diablo 2\binkw32.dll (v3.0.0.0)
  Diablo 2\d2fps.dll
Detected monitor fps: 165
```

I've however now noticed a problem with the menu FPS fix (see `Failed to apply patch at: D2Win.dll+0x18a4e` above). As far as I remember, the file I've got here is the unmodified original from the v1.13c patch.

**Comment by Jarcho** [comment]
> It was more like a full second, and then the screen came back to normal. I was probably lucky not to experience it before then 🙂.

A long time like that I haven't seen before. Might be a new thing.

> It does detect the refresh rate properly.

That is good to know.

> I've however now noticed a problem with the menu FPS fix (see Failed to apply patch at: D2Win.dll+0x18a4e above). As far as I remember, the file I've got here is the unmodified original from the v1.13c patch.

SHA-256 for 1.13c d2win.dll is `a9afb52d5116f77534f3e83eeb71e6d2d980a6b1c59be5dd38d503c2334730fb` in case you want to check. Also possible that something else is patching the code first. I think BaseMod includes a patch for this and it might be loaded before d2fps is.

I'll add an option to bypass the integrity check. If the game blows up afterwards then that's what ends up happening.

> I'd also like to ask whether it wouldn't be a problem for you to compile a new version with the music fix. I want to test this new motion predictor more but having to re-enable music after every Alt+Tab is too cumbersome, so I've restored the old 0.100.0.0 version for now.

Will do.

**Comment by tomasz1986** [comment]
> SHA-256 for 1.13c d2win.dll is `a9afb52d5116f77534f3e83eeb71e6d2d980a6b1c59be5dd38d503c2334730fb` in case you want to check. Also possible that something else is patching the code first. I think BaseMod includes a patch for this and it might be loaded before d2fps is.

The hash matches, so it's likely BaseMod. I tried to disable the integrity check and the game seemed to work properly, however on launch it sometimes got very stuttery and laggy which lasted for about 10 seconds or so. Because of this, I've re-enabled the integrity check as its better to have the menu play in 25 fps than to have this laggy state on start.

> Main post updated with the new build.

I've been testing it today, and I've just had a strange crash when trying to minimise the game. However, it wasn't just the game that crashed but the whole graphics driver which then managed to recover itself automatically.

```
Game version: 1.0.13.60 (supported)

Apparent Windows version: 10.0 (build 19044).
Actual Windows version: 10.0 (build 0).
Suggesting game size 1280x720.
Initializing D2FPS.
Using 'Interval1' sync strategy.
Using 'FlipDiscard' swap strategy.
Created device supports D3D_FEATURE_LEVEL_11_1.
Using 'ResizeBuffers' backbuffer sizing strategy.
Device context supports ID3D11DeviceContext1. Will use this to discard resources and views.
Setting maximum frame latency to 1.
The device supports 2048 textures per atlas.
Total size of texture caches is 240256 kB.
Using DXGI_FORMAT_R10G10B10A2_UNORM for the render buffer.
Using DXGI_FORMAT_R10G10B10A2_UNORM for the render buffer.
Sizes: desktop 2560x1440, window 2400x1800, game 800x600, render 1920x1440
Using DXGI_FORMAT_R10G10B10A2_UNORM for the render buffer.
Sizes: desktop 2560x1440, window 2400x1800, game 640x480, render 1920x1440
Using DXGI_FORMAT_R10G10B10A2_UNORM for the render buffer.
Sizes: desktop 2560x1440, window 2400x1800, game 800x600, render 1920x1440
d2dx::RenderContext::Present line 489
HRESULT: 0x0019f730
Message: unknown error
```

**Comment by tomasz1986** [comment]
I've just experienced this small glitch today.

![IMG_20230825_220722](https://github.com/Jarcho/d2dx/assets/5626656/d3fb63dd-6da3-4926-a56b-5573eba68c93)

Not sure what actually happened but the Windows mouse cursor got stuck on top of the in-game mouse cursor. This likely happened when Alt+Tabing out of and back into the game. Once stuck like that, I tried changing to window mode and back to full screen, changing the in-game resolution, etc. but I still couldn't get it fixed, so in the end I had to quit and restart the game.

I'm attaching logs just in case although I don't think there is really anything useful in them.

[d2dx_log.txt](https://github.com/Jarcho/d2dx/files/12443257/d2dx_log.txt)
[d2fps.log.txt](https://github.com/Jarcho/d2dx/files/12443258/d2fps.log.txt)
[D2230825.txt](https://github.com/Jarcho/d2dx/files/12443259/D2230825.txt)

**Comment by Jarcho** [comment]
I'm guessing one of the other mods messed the `ShowCursor` count up somehow. Not sure exactly how that would happen, but I can't think of anything else that would cause it.

Next build will have it at least be fixable by moving the mouse into the title bar and back.

**Comment by Jarcho** [comment]
This has been released as v0.101.0. The released version is basically the same, except it will patch over BaseMod's menu cpu-fix patch  (not it's in-game fps patch, that one will need to be disabled in `basemod.ini`). Several other fps/cpu-fix patches will also be patched over as well.
